Support groups in DenseConvDims

DhairyaLGandhi opened the pull request:
Also propagate it to the kernel calls.

CarloLucibello commented:
Our cpu conv kernels don't seem to support groups, do they?

CarloLucibello commented:
> Our cpu conv kernels don't seem to support groups, do they?

@DhairyaLGandhi can I get any answers here? Which conv implementation supports groups? Can you point to some specific line of code?

CarloLucibello commented:
ah you are implementing groups support right now, don't have to be that mysterious :)

DhairyaLGandhi commented:
So https://github.com/FluxML/NNlib.jl/pull/289/commits/115784b8128d0e5c4cbdd0a1595328aa40adc862 is a "working" version which is very memory wasteful, but I have it here right now in case someone was eager to see it working in NNlib. I will be moving to the non-allocating version of this soon. The issue is Julia doesn't like to look into the `stride` of deep and lazy wrapper types around `SubArray`, `ReshapedArray` etc and can cause races, or return garbage. In the previous version, we had segfaults because the `DenseConvDims` needs to be extra careful about the encoding dims that include the groups.

DhairyaLGandhi commented:
For the record, not using tasks is as fast as the original version for the regular convolutions (ie with `groups = 1`), and maybe the serial version is faster anyway.

DhairyaLGandhi commented on src/conv.jl:
```diff
@@ -188,7 +189,17 @@ for (front_name, backend) in (
+                                       channels_in(cdims) ÷ groupcount(cdims))
+            w_cs = Iterators.partition(1:size(in2, 5),
+                                       channels_out(cdims) ÷ groupcount(cdims))
+            cdims2 = DenseConvDims(cdims, G = 1, C_in = channels_in(cdims) ÷ groupcount(cdims), C_out = channels_out(cdims) ÷ groupcount(cdims))
```
@staticfloat  I'm wondering if its acceptable to modify the `cdims`, here. We can't use `DenseConvDims` of course, but I have an overload for `groupcount(::ConvDims)` for the default case anyway.

DhairyaLGandhi commented:
Oh this is satisfying. Starting Julia with `-t2` on my tiny MacBook Pro with a quad core. For context this results in about a 20% reduction compared to doing the sequential thing. It might be safer to not do that, but RFC?
```Julia
julia> @btime y = conv($(rand(Float32, 28, 28, 100, 2)), $(rand(Float32, 3, 3, 20, 15)), groups = 5);
  947.722 μs (159 allocations: 4.81 MiB)
```

Either way we now have the system in place to do pass it to CUDNN. Should I also add a stub to `DepthwiseConvDims` for consistency? I might do that actually.

CarloLucibello commented on src/conv.jl:
```diff
@@ -188,7 +189,20 @@ for (front_name, backend) in (
                         out::AbstractArray{T,5}, in1::AbstractArray{T,5},
                         in2::AbstractArray{T,5}, cdims::ConvDims; kwargs...) where {T <: $G}
-            $(Symbol("$(front_name)_$(backend)!"))(out, in1, in2, cdims; kwargs...)
+            x_cs = Iterators.partition(1:size(in1, 4),
+                                       channels_in(cdims) ÷ groupcount(cdims))
```
```suggestion
            x_cs = Iterators.partition(1:channels_in(cdims) , groupcount(cdims))
```
the number of partitions should be equal to the number of groups, right?

CarloLucibello replied:
ops, forget what i said, the second argument of partition is the number of element in each partition

CarloLucibello commented on src/conv.jl:
```diff
@@ -188,7 +189,20 @@ for (front_name, backend) in (
+                                   G = 1,
+                                   C_in = channels_in(cdims) ÷ groupcount(cdims),
+                                   C_out = channels_out(cdims) ÷ groupcount(cdims))
+            Threads.@sync Threads.@spawn for (xc, wc) in zip(x_cs, w_cs)
```
```suggestion
            Threads.@threads for (xc, wc) in zip(x_cs, w_cs)
```
are these the same?

DhairyaLGandhi replied:
They are different. We are looking to keep the threads as free as possible.

CarloLucibello replied:
> They are different

Yes I checked the docs and saw they are different. You are spawning a single thread carrying over the whole for loop, so in practice this the same as doing it single-threaded, right?

The correct way to parallelize this loop seems to be with `@threads`.

> We are looking to keep the threads as free as possible

I don't know what that means

DhairyaLGandhi replied:
We want to spawn tasks that perform convolutions, which are in turn threaded.

CarloLucibello replied:
I don't know julia's multithreading model very well, so what I'm saying could be wrong but still I give my two cents. 
How things are currently done in this PR is that you spawn a single task running the whole loop, so groups are convolved sequentially. Each group's convolution is then parallelized. I think a better approach is to apply `@threads` to the for loop instead, so that each iteration becomes a taks. Then  tasks and tasks-within-tasks are automatically scheduled by julia.

CarloLucibello commented:
>  For context this results in about a 20% reduction compared to doing the sequential thing. It might be safer to not do that, but RFC?

I don't see any safety issue in going multithreaded here, computations are independent.

We should check that using threads doesn't have a performance impact on the standard convolution case (G=1).

DhairyaLGandhi commented:
Its a question of computational complexity. The existing cases aren't affected much at all.

CarloLucibello commented:
I think the existing cases are affected because spawning threads doesn't come for free, but hopefully could be negligible compared to the convolution cost. If that's not the case, we can still dispatch on the G=1 case.

In any case, we should check we don't have performance regressions in standard convolution before merging
(+ the obvious stuff: existing tests are not passing, new tests are needed, docstring should be updated)

DhairyaLGandhi commented:
Way ahead of you :)

darsnack commented:
> For context this results in about a 20% reduction compared to doing the sequential thing.

🎉

pxl-th commented:
Hi! Just curious if there is any update to this?
Really would like to see `DepthwiseConv` working on gpu again (as it is not working with the latest Flux version).
If you need any help I'm happy to help :)

DhairyaLGandhi commented:
Yes! I had also altered the backwards passes to use the correct group count but haven't pushed them out yet. I'll do that when I'm on my computer next. Apart from that, we need to thread this to cuda, and add tests.

pxl-th commented:
Great! If you want, I can work on NNlibCUDA.jl to account for changes in `ConvDims` and add support for groups and open PR there once this PR is merged.

DhairyaLGandhi commented:
That would be very appreciated. I have the changes all done here to get things running now. I didn't expect needing to write out the data and filter backwards passes manually though, so some cleanup may be in order. But overall, it should work fine.

staticfloat commented on src/dim_helpers/DenseConvDims.jl:
```diff
@@ -5,35 +5,47 @@ export DenseConvDims
+    end
+   
+    # Ensure groups make sense
+    # if x_size[end-1] ÷ groups == w_size[end] && 
+    # @show x_size, w_size
```
What is this supposed to be?

staticfloat commented on src/dim_helpers/DenseConvDims.jl:
```diff
@@ -56,22 +68,26 @@ end
 
+    # Check the groups match
+    @assert channels_in(cdims) % groupcount(cdims) == 0 DimensionMismatch("Groups ($(groupcount(cdims))) should be divisble by input channels $(channels_in(cdims))")
+    # @assert channels_out(cdims) % groupcount(cdims) == 0 DimensionMismatch("Groups ($(groupcount(cdims))) should be divisble by input channels $(channels_out(cdims))")
```
Is this supposed to be something?

CarloLucibello commented on src/conv.jl:
```diff
@@ -172,8 +176,94 @@ end
+    )
+
+    # These are the GEMM types we will accelerate with `im2col`
+    G = Union{[x[2] for x in gemm_datatype_mappings]...}
```
this is already defined down below

CarloLucibello commented on src/dim_helpers/DenseConvDims.jl:
```diff
@@ -5,35 +5,45 @@ export DenseConvDims
         M - 2,
         w_size[1:end-2],
         x_size[end-1],
         w_size[end],
```
```suggestion
        w_size[end] * groups,
```
`channels_out` should give the total number of output channels (consistently with `channels_in`.

After modifying this, some of the size check in the PR should be adapted

CarloLucibello commented on src/dim_helpers/DenseConvDims.jl:
```diff
@@ -56,22 +66,26 @@ end
     @assert w[1:M-2] == kernel_size(cdims) DimensionMismatch("Kernel spatial size ($(w[1:M-2]) vs. $(kernel_size(cdims)))")
 
+    # Check the groups match
+    @assert channels_in(cdims) % groupcount(cdims) == 0 DimensionMismatch("Groups ($(groupcount(cdims))) should be divisble by input channels $(channels_in(cdims))")
```
this assert is redundat with the ones above, can be removed

DhairyaLGandhi commented:
Addressed all the comments. I've also added more tests and docs. @pxl-th could you rebase your NNlibCUDA PR? Any other comments?

CarloLucibello commented:
not sure this respecting the fact that `channels_in(cdims) == size(x, 3)` and `channels_out(cdims) == size(y, 3)`

CarloLucibello commented on src/conv.jl:
```diff
@@ -172,8 +176,94 @@ end
 for (front_name, backend) in (
         # This maps from public, front-facing name, to internal backend name
         :conv                   => :im2col,
-        :∇conv_data             => :im2col,
-        :∇conv_filter           => :im2col,
+        # :∇conv_data             => :im2col,
+        # :∇conv_filter           => :im2col,
+    )
```
we can eliminate the loop and define explicitly the `conv` backend as well

CarloLucibello commented on src/conv.jl:
```diff
@@ -169,25 +173,106 @@ end
+              Threads.@spawn $(Symbol("$(front_name)_$(backend)!"))(y, x, w, cdims2; kwargs...)
+            end
+
+           return out
```
we use 4 space indent here

CarloLucibello commented on src/conv.jl:
```diff
@@ -169,25 +173,106 @@ end
+                         C_in = channels_in(cdims) ÷ groupcount(cdims),
+                         C_out = channels_out(cdims) ÷ groupcount(cdims))
+
+    Threads.@sync for (xc, yc, wc) in zip(dx_cs, dy_cs, w_cs)
+      dxv = @view out[ntuple(i -> i == 4 ? xc : Colon(), 5)...]
+      dyv = @view in1[ntuple(i -> i == 4 ? yc : Colon(), 5)...]
+      wv = @view in2[ntuple(i -> i == 5  ? wc : Colon(), 5)...]
+      Threads.@spawn ∇conv_data_im2col!(dxv, dyv, wv, cdims2; kwargs...)
+    end
```
does the spawn approach allow for composable threading? Ideally each `conv` call on the slices should be multithreaded as well if the machine has enough threads

DhairyaLGandhi replied:
This is what I see, which suggests the impact on the G = 1 case is tractable. I have answered this in a [regular comment](https://github.com/FluxML/NNlib.jl/pull/289#issuecomment-880602522) since GitHub didn't show me the reply pane for some reason.

```Julia
julia> x′ = rand(Float32, 28, 28, 20, 2);

julia> w′ = rand(Float32, 3, 3, 20, 15);

julia> cdims = DenseConvDims(x′, w′);

julia> @btime conv($x′, $w′, $cdims);
  230.201 μs (23 allocations: 555.95 KiB) # master; -t1
  243.761 μs (50 allocations: 557.77 KiB) # PR; -t1

  205.723 μs (28 allocations: 1.01 MiB) # master; -t2
  212.537 μs (47 allocations: 1.01 MiB)  # PR; -t2

julia> w′ = rand(Float32, 3, 3, 3, 6);

julia> x′ = rand(Float32, 28, 28, 3, 2);

julia> cdims = DenseConvDims(x′, w′);

julia> @btime conv($x′, $w′, $cdims);
  47.205 μs (23 allocations: 104.39 KiB) # master; -t1
  55.600 μs (50 allocations: 106.20 KiB) # PR; -t1

  35.217 μs (27 allocations: 176.05 KiB) # master; -t2
  40.850 μs (47 allocations: 177.14 KiB) # PR; -t2
```

CarloLucibello replied:
A little regression, but definitely worth it for this important feature. It will only affect cpu conv, so no big deal

DhairyaLGandhi commented:
Done, also made `channels_out` respect groups. I think storing the groups is needed anyway along with the regular `C_out` (since we ultimately use the G = 1 definition).

DhairyaLGandhi commented:
Ah, almost forgot - we need `channels_out` to mean the same as master - since that data is needed. See `∇conv_data!`

DhairyaLGandhi commented:
I've addressed the reviews - thanks @staticfloat! I had a couple of debugging steps that I've cleaned up now. Also comes with more tests, and docs!

Re the added check - its there so we can give a clearer message about groups config.

CarloLucibello commented on test/conv.jl:
```diff
@@ -648,6 +649,33 @@ else
+   end
+
+   # Currently hangs due to a FiniteDifferences issue
+   @test_skip gradtest((x, w) -> sum(conv(x, w, cdims)), x′, w′)
```
you can try with 
```suggestion
   @test_skip gradtest((x, w) -> sum(conv(x, w, cdims)), x′, w′, fdm=:forward)
```
or  `fdm=:backward` and see if this passes

DhairyaLGandhi replied:
same effect, although this did reach `conv!`
```Julia
Grouped Convolutions: Error During Test at REPL[10]:1
  Got exception outside of a @test
  TaskFailedException

      nested task error: InterruptException:
Stacktrace:
  [1] try_yieldto(undo::typeof(Base.ensure_rescheduled))
    @ Base ./task.jl:705
  [2] wait
    @ ./task.jl:764 [inlined]
  [3] wait(c::Base.GenericCondition{Base.Threads.SpinLock})
    @ Base ./condition.jl:106
  [4] _wait(t::Task)
    @ Base ./task.jl:288
  [5] sync_end(c::Channel{Any})
    @ Base ./task.jl:344
  [6] macro expansion
    @ ./task.jl:383 [inlined]
  [7] conv!(out::Array{Float32, 5}, in1::Array{Float32, 5}, in2::Array{Float32, 5}, cdims::DenseConvDims{3, (3, 3, 1), 100, 15, 5, (1, 1, 1), (0, 0, 0, 0, 0, 0), (1, 1, 1), false}; kwargs::Base.Iterators.Pairs{Union{}, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
    @ NNlib ~/Downloads/forks/NNlib.jl/src/conv.jl:201
  [8] conv!(out::Array{Float32, 5}, in1::Array{Float32, 5}, in2::Array{Float32, 5}, cdims::DenseConvDims{3, (3, 3, 1), 100, 15, 5, (1, 1, 1), (0, 0, 0, 0, 0, 0), (1, 1, 1), false})
    @ NNlib ~/Downloads/forks/NNlib.jl/src/conv.jl:192
```

and some times I get:
```
fatal: error thrown and no exception handler available.
InterruptException()
sigatomic_end at ./c.jl:437 [inlined]
task_done_hook at ./task.jl:487
jl_apply_generic at /Applications/Julia-1.6.app/Contents/Resources/julia/lib/julia/libjulia-internal.1.6.dylib (unknown line)
jl_finish_task at /Applications/Julia-1.6.app/Contents/Resources/julia/lib/julia/libjulia-internal.1.6.dylib (unknown line)
start_task at /Applications/Julia-1.6.app/Contents/Resources/julia/lib/julia/libjulia-internal.1.6.dylib (unknown line)
```

The other tests haven't complained though.

CarloLucibello replied:
Will fix later. We generally need more robust tests

DhairyaLGandhi commented:
@pxl-th do you think you can test your NNlibCUDA PR off of this one? The API is mostly the same, but good to get a check that we are passing into CUDNN properly.

DhairyaLGandhi commented:
Let's merge this. I'll wait till tomorrow, so if anyone has final thoughts we can address those.

pxl-th commented:
@DhairyaLGandhi I've tested it. Everything seems to work, all the tests are passing.

DhairyaLGandhi commented:
Nice, let's get these PRs merged then.
